### PR TITLE
Rholang: ADT: Add locally free bitset.

### PR DESCRIPTION
**Pull request by guardbotmk3**
This will allow us to do matches of terms with environments. It turns out that a
free variable is not the same as true, because you can only bind things that
don't have any locally free variables. For a fully substituted term, all we need
is the lowest value used in a term to figure out if we can match, but with an
environment, we need to know what the lowest would be once all the variables in
the environment are substituted. Hence, a bitset.

This will likely complicate using a protobuf. On the plus side, It's info that
could be re-calculated on deserialize, or we can serialize the bitset as bytes
in the proto.

### CODE REVIEW

**Comment by guardbotmk3** [comment]
OK, I made this work with match. Please take a look.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala [hunk 1]
```diff
@@ -143,7 +146,9 @@ object ProcNormalizeMatcher {
       val desugaredIf = Match(
         targetResult.par,
         List((GBool(true), trueCaseBody.par), (GBool(false), falseCaseBody.par)),
-        freeCount)
+        freeCount,
+        trueCaseBody.par.locallyFree | falseCaseBody.par.locallyFree
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: `targetResult.par.locallyFree | trueCaseBody.par.locallyFree | falseCaseBody.par.locallyFree`

Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/normalize.scala]: Good catch. Thank you.

**Comment by guardbotmk3** [comment]
OK, I handled your comment and merged to take care of conflicts with Joe's substitute code.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 2]
```diff
@@ -111,39 +117,26 @@ object Substitute {
...
+      case EMap(kvs, freeCount, locallyFree) =>
         val _ps = for { (p1, p2) <- kvs } yield {
           (substitute(p1), substitute(p2))
         }
-        val fc = (0 /: _ps) {
-          case (i, (p1, p2)) =>
-            i + p1.freeCount + p2.freeCount
-        }
-        EMap(_ps, fc)
+        EMap(_ps, freeCount, locallyFree)
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Do we not need a similar pattern here `locallyFree.from(env.level).map(x => x - env.level)`?

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala [hunk 3]
```diff
@@ -55,7 +55,8 @@ object Substitute {
       Nil,
       for { m <- term.matches } yield substitute(m),
       term.id,
-      term.freeCount
+      term.freeCount,
+      term.locallyFree.from(env.level).map(x => x - env.level)
```
Comment (KentShikama) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Why is env.level defined as

`def level: Int = env.keys.max + 1`

and not just

`def level: Int = env.keys.max`

Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Well I guess the definition of env.level is fine but don't we have to do

`term.locallyFree.from(env.level - 1).map(x => x - env.level + 1)`

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 2 above)
Comment (KentShikama) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: See related comment below for potential change to this pattern.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 3 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: No. If the max key in the env is 0, then we need to shift the environment by 1.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala (hunk 2 above)
Comment (guardbotmk3) [reply rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala]: Good catch.

**Comment by guardbotmk3** [comment]
OK, Please look again.

**Pull request merged**